File: anchor/programs/core-bridge/src/legacy/processor/governance/upgrade_contract.rs
use crate::{
    constants::UPGRADE_SEED_PREFIX,
    error::CoreBridgeError,
    legacy::instruction::EmptyArgs,
    state::{BridgeProgramData, Claim, PartialPostedVaaV1, VaaV1MessageHash},
    utils::GOVERNANCE_DECREE_START,
};
use anchor_lang::prelude::*;
use solana_program::{bpf_loader_upgradeable, program::invoke_signed};
use wormhole_io::Readable;
use wormhole_solana_common::{BpfLoaderUpgradeable, SeedPrefix};

const ACTION_CONTRACT_UPGRADE: u8 = 1;

#[derive(Accounts)]
pub struct UpgradeContract<'info> {
    #[account(mut)]
    payer: Signer<'info>,

    #[account(
        mut,
        seeds = [BridgeProgramData::seed_prefix()],
        bump,
    )]
    bridge: Account<'info, BridgeProgramData>,

    #[account(
        seeds = [
            PartialPostedVaaV1::seed_prefix(),
            posted_vaa.try_message_hash()?.as_ref()
        ],
        bump
    )]
    posted_vaa: Account<'info, PartialPostedVaaV1>,

    #[account(
        init,
        payer = payer,
        space = Claim::INIT_SPACE,
        seeds = [
            posted_vaa.emitter_address.as_ref(),
            &posted_vaa.emitter_chain.to_be_bytes(),
            &posted_vaa.sequence.to_be_bytes()
        ],
        bump,
    )]
    claim: Account<'info, Claim>,

    /// CHECK: We need this upgrade authority to invoke the BPF Loader Upgradeable program to
    /// upgrade this program's executable.
    #[account(
        seeds = [UPGRADE_SEED_PREFIX],
        bump,
    )]
    upgrade_authority: AccountInfo<'info>,

    /// CHECK: This account is needed for the BPF Loader Upgradeable program.
    spill: UncheckedAccount<'info>,

    /// CHECK: This account is needed for the BPF Loader Upgradeable program.
    ///
    /// NOTE: This account's pubkey is what is encoded in the governance VAA. We check this in the
    /// instruction handler.
    buffer: UncheckedAccount<'info>,

    /// CHECK: This account is needed for the BPF Loader Upgradeable program.
    program_data: UncheckedAccount<'info>,

    /// CHECK: Unnecessary account.
    _this_program: UncheckedAccount<'info>,

    /// CHECK: Previously needed sysvar.
    _rent: UncheckedAccount<'info>,

    /// CHECK: Previously needed sysvar.
    _clock: UncheckedAccount<'info>,

    bpf_loader_upgradeable_program: Program<'info, BpfLoaderUpgradeable>,
    system_program: Program<'info, System>,
}

impl<'info> UpgradeContract<'info> {
    fn accounts(ctx: &Context<Self>) -> Result<()> {
        let action = crate::utils::require_valid_governance_posted_vaa(
            &ctx.accounts.posted_vaa,
            &ctx.accounts.bridge,
        )?;

        require_eq!(
            action,
            ACTION_CONTRACT_UPGRADE,
            CoreBridgeError::InvalidGovernanceAction
        );

        let acc_info: &AccountInfo = ctx.accounts.posted_vaa.as_ref();
        let mut data = &acc_info.data.borrow()[GOVERNANCE_DECREE_START..];

        // Read the implementation pubkey and check against the buffer in our account context.
        require_keys_eq!(
            Pubkey::new_from_array(Readable::read(&mut data).unwrap()),
            ctx.accounts.buffer.key()
        );

        // Done.
        Ok(())
    }
}

#[access_control(UpgradeContract::accounts(&ctx))]
pub fn upgrade_contract(ctx: Context<UpgradeContract>, _args: EmptyArgs) -> Result<()> {
    // Mark the claim as complete.
    ctx.accounts.claim.is_complete = true;

    // Finally upgrade.
    invoke_signed(
        &bpf_loader_upgradeable::upgrade(
            &crate::ID,
            &ctx.accounts.buffer.key(),
            &ctx.accounts.upgrade_authority.key(),
            &ctx.accounts.spill.key(),
        ),
        &ctx.accounts.to_account_infos(),
        &[&[UPGRADE_SEED_PREFIX, &[ctx.bumps["upgrade_authority"]]]],
    )
    .map_err(Into::into)
}

File: solana/tests/helpers/coreBridge/legacy/instructions/governance/upgradeContract.ts
import { ParsedVaa } from "@certusone/wormhole-sdk";
import { BN } from "@coral-xyz/anchor";
import {
  AccountMeta,
  PublicKey,
  SYSVAR_CLOCK_PUBKEY,
  SYSVAR_RENT_PUBKEY,
  SystemProgram,
  TransactionInstruction,
} from "@solana/web3.js";
import { CoreBridgeProgram } from "../../..";
import { BPF_LOADER_UPGRADEABLE_PROGRAM_ID, ProgramData } from "../../../../native";
import { BridgeProgramData, Claim, PostedVaaV1, upgradeAuthorityPda } from "../../state";

export type LegacyUpgradeContractContext = {
  payer: PublicKey;
  bridge?: PublicKey;
  postedVaa?: PublicKey;
  claim?: PublicKey;
  upgradeAuthority?: PublicKey;
  spill?: PublicKey;
  buffer?: PublicKey;
  programData?: PublicKey;
  thisProgram?: PublicKey;
  rent?: PublicKey;
  clock?: PublicKey;
  bpfLoaderUpgradeableProgram?: PublicKey;
};

export function legacyUpgradeContractIx(
  program: CoreBridgeProgram,
  accounts: LegacyUpgradeContractContext,
  parsed: ParsedVaa
) {
  const programId = program.programId;
  const { emitterChain, emitterAddress, sequence, hash } = parsed;

  let {
    payer,
    bridge,
    postedVaa,
    claim,
    upgradeAuthority,
    spill,
    buffer,
    programData,
    thisProgram,
    rent,
    clock,
    bpfLoaderUpgradeableProgram,
  } = accounts;

  if (bridge === undefined) {
    bridge = BridgeProgramData.address(programId);
  }

  if (postedVaa === undefined) {
    postedVaa = PostedVaaV1.address(programId, Array.from(hash));
  }

  if (claim === undefined) {
    claim = Claim.address(
      programId,
      Array.from(emitterAddress),
      emitterChain,
      new BN(sequence.toString())
    );
  }

  if (upgradeAuthority === undefined) {
    upgradeAuthority = upgradeAuthorityPda(programId);
  }

  if (spill === undefined) {
    spill = payer;
  }

  if (buffer === undefined) {
    buffer = new PublicKey(parsed.payload.subarray(-32));
  }

  if (programData === undefined) {
    programData = ProgramData.address(programId);
  }

  if (thisProgram === undefined) {
    thisProgram = programId;
  }

  if (rent === undefined) {
    rent = SYSVAR_RENT_PUBKEY;
  }

  if (clock === undefined) {
    clock = SYSVAR_CLOCK_PUBKEY;
  }

  if (bpfLoaderUpgradeableProgram === undefined) {
    bpfLoaderUpgradeableProgram = BPF_LOADER_UPGRADEABLE_PROGRAM_ID;
  }

  const keys: AccountMeta[] = [
    {
      pubkey: payer,
      isWritable: true,
      isSigner: true,
    },
    {
      pubkey: bridge,
      isWritable: true,
      isSigner: false,
    },
    {
      pubkey: postedVaa,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: claim,
      isWritable: true,
      isSigner: false,
    },
    {
      pubkey: upgradeAuthority,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: spill,
      isWritable: true,
      isSigner: false,
    },
    {
      pubkey: buffer,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: programData,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: thisProgram,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: rent,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: clock,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: bpfLoaderUpgradeableProgram,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: SystemProgram.programId,
      isWritable: false,
      isSigner: false,
    },
  ];
  const data = Buffer.alloc(1, 5);

  return new TransactionInstruction({
    keys,
    programId,
    data,
  });
}
